### PR TITLE
[Sharding] Shard aware updates for internal API

### PR DESCRIPTION
**Pull request by agourlay**
This PR adds a shard aware update API. https://github.com/qdrant/qdrant/issues/419

A new optional `shard_selection` parameter is propagated down the stack to choose how to handle the update:
- `update_from_client` is the current public API which splits batches by shards for redistribution
- `update_from_peer` is the new API which expects the `shard_selection` to be one of the peer's local shard (the error handling needs to be revised) 

In order to validate the current direction I have implemented a single method on a newly introduced internal gRPC API to get a feel for the complete integration.

I tried to mitigate code duplication by introducing a common API layer `points_common.rs`.

### CODE REVIEW

File: lib/collection/src/lib.rs [hunk 1]
```diff
@@ -205,6 +206,43 @@ impl Collection {
...
+        wait: bool,
+    ) -> CollectionResult<UpdateResult> {
+        match shard_selection {
+            Some(shard_selection) => {
+                self.update_from_peer(operation, shard_selection, wait)
+                    .await
+            }
+            None => self.update_from_client(operation, wait).await,
+        }
+    }
```
Comment (e-ivkov) [inline lib/collection/src/lib.rs:221]: Do we need a public general purpose `update` if we have public `update_from_client` and `update_from_peer`?

Comment (agourlay) [reply lib/collection/src/lib.rs:221]: I think you are making a good point, I will see how it looks like without the branching on the `Option` :+1:

Comment (agourlay) [reply lib/collection/src/lib.rs]: I have removed the general purpose update API at the collection level, now the branching takes place one level higher in the `TableOfContent`.

https://github.com/qdrant/qdrant/pull/427/commits/68633543ad7a8354e1464c703e7675c1cec8388f

Is it what you had in mind? I am not sure it is possible to push the branching higher up without code duplication.

Comment (e-ivkov) [reply lib/collection/src/lib.rs]: I actually thought to push it all the way to tonic API as there we can see that something comes from peer and so we call the appropriate method. But if it leads to a lot of code duplication then I don't mind have it as it is.

File: src/tonic/api/points_internal_api.rs [hunk 2]
```diff
@@ -0,0 +1,44 @@
+use tonic::{Request, Response, Status};
+
+use crate::tonic::api::points_common::upsert;
+use api::grpc::qdrant::points_internal_server::PointsInternal;
+use api::grpc::qdrant::{PointsOperationResponse, UpsertPointsInternal};
+use std::sync::Arc;
+use storage::content_manager::toc::TableOfContent;
+
+pub struct PointsInternalService {
```
Comment (e-ivkov) [inline src/tonic/api/points_internal_api.rs:9]: Maybe a doc comment that this is intended for P2P communication between peers as part of remote shard functionality?

File: lib/api/src/grpc/proto/points_internal_service.proto [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+import "google/protobuf/struct.proto";
+
+service PointsInternal {
+  rpc Upsert (UpsertPointsInternal) returns (PointsOperationResponse) {}
+}
+
+message UpsertPointsInternal {
+  UpsertPoints upsert_points = 1;
+  uint32 shard_id = 2;
+}
```
Comment (e-ivkov) [inline lib/api/src/grpc/proto/points_internal_service.proto:20]: Please add an empty line at the end of file

File: src/tonic/mod.rs [hunk 4]
```diff
@@ -44,13 +46,15 @@ pub fn init(toc: Arc<TableOfContent>, host: String, grpc_port: u16) -> std::io::
...
             let points_service = PointsService::new(toc.clone());
+            let points_internal_service = PointsInternalService::new(toc.clone());
 
             log::info!("Qdrant gRPC listening on {}", grpc_port);
 
             Server::builder()
                 .add_service(QdrantServer::new(service))
                 .add_service(CollectionsServer::new(collections_service))
                 .add_service(PointsServer::new(points_service))
+                .add_service(PointsInternalServer::new(points_internal_service)) // TODO serve from different port
```
Comment (e-ivkov) [inline src/tonic/mod.rs:57]: Maybe an issue for this?

File: lib/api/src/grpc/proto/points_internal_service.proto (hunk 3 above)
Comment (e-ivkov) [reply lib/api/src/grpc/proto/points_internal_service.proto:20]: Github does not like it without empty line :)

File: src/tonic/mod.rs (hunk 4 above)
Comment (agourlay) [reply src/tonic/mod.rs:57]: Good point :+1: 

https://github.com/qdrant/qdrant/issues/430

**Review by e-ivkov — APPROVED** [review]

**Comment by agourlay** [comment]
I will explore if it is possible to create a **single internal entry point** for the collection update operations by having as input argument:

```rust
pub struct CollectionShardUpdateOperations {
    collection_update_operation: CollectionUpdateOperations,
    shard_id: ShardId,
}
```

The update pipeline and remote shards are perfectly aware of `CollectionUpdateOperations` so it would make the invocation and propagation process really smooth internally.

I hope it won't be too much boilerplate to create a proto. definition for `CollectionUpdateOperations` that will hopefully reuses some existing proto. definition.

**Comment by agourlay** [comment]
Follow up on the experimentation for having a single internal entry point mentioned above.

While the idea is very attractive, it does not work out in practice unfortunately.

To see why let's start by defining the proto. definition for the message.

```proto
message CollectionShardUpdateOperations {
  oneof operation_one_of {
    PointOperation point_operation = 1;
    PayloadOperation payload_operation = 2;
    FieldIndexOperation field_operation = 3;
  }
  uint32 shard_id = 4;
}
```

For `PointOperation` we could have something reusing existing definitions

```proto
message PointOperation {
  oneof operation_one_of {
    UpsertPoints upsert_points = 1;
    DeletePoints delete_points = 2;
    DeletePointsByFilter delete_points_by_filter = 3;
  }
}
```

Unfortunately it does not work because the existing definition of `UpsertPoints` describe an RPC call and not solely the data transferred.

```proto
message UpsertPoints {
  string collection_name = 1; // name of the collection
  optional bool wait = 2; // Wait until the changes have been applied?
  repeated PointStruct points = 3;
}
```

This means that this approach would need extra work to work:
- a complete in-depth redefinition in proto. of the `PointOperation`, `PayloadOperation` & `FieldIndexOperation` enums
- new two-ways conversions implementation between grpc amd domain types

The bottom line is that although it is very desirable to get a single entry points for collection update operations using the `CollectionUpdateOperations`, it comes at a high cost in boilerplate.

I recommend going with the multi-endpoint approach for the time being.

**Review by generall — APPROVED** [review]
